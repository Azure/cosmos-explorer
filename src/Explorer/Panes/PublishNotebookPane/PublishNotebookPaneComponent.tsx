import { Dropdown, IDropdownProps, ITextFieldProps, Stack, Text, TextField } from "@fluentui/react";
import { ImmutableNotebook } from "@nteract/commutable";
import React, { FunctionComponent, useState } from "react";
import { GalleryCardComponent } from "../../Controls/NotebookGallery/Cards/GalleryCardComponent";
import * as FileSystemUtil from "../../Notebook/FileSystemUtil";
import { SnapshotRequest } from "../../Notebook/NotebookComponent/types";
import { NotebookUtil } from "../../Notebook/NotebookUtil";
import "./styled.less";

export interface PublishNotebookPaneProps {
  notebookName: string;
  notebookAuthor: string;
  notebookTags: string;
  notebookDescription: string;
  notebookCreatedDate: string;
  notebookObject: ImmutableNotebook;
  notebookContentRef: string;
  imageSrc: string;

  onError: (formError: string, formErrorDetail: string, area: string) => void;
  clearFormError: () => void;
  setNotebookName: (newValue: string) => void;
  setNotebookDescription: (newValue: string) => void;
  setNotebookTags: (newValue: string) => void;
  setImageSrc: (newValue: string) => void;
  onTakeSnapshot: (request: SnapshotRequest) => void;
}

enum ImageTypes {
  Url = "URL",
  CustomImage = "Custom Image",
  TakeScreenshot = "Take Screenshot",
  UseFirstDisplayOutput = "Use First Display Output",
}

export const PublishNotebookPaneComponent: FunctionComponent<PublishNotebookPaneProps> = ({
  notebookName,
  notebookTags,
  notebookDescription,
  notebookAuthor,
  notebookCreatedDate,
  notebookObject,
  notebookContentRef,
  imageSrc,
  onError,
  clearFormError,
  setNotebookName,
  setNotebookDescription,
  setNotebookTags,
  setImageSrc,
  onTakeSnapshot,
}: PublishNotebookPaneProps) => {
  const [type, setType] = useState<string>(ImageTypes.CustomImage);
  const CARD_WIDTH = 256;
  const cardImageHeight = 144;
  const cardHeightToWidthRatio = cardImageHeight / CARD_WIDTH;

  const maxImageSizeInMib = 1.5;

  const descriptionPara1 =
    "When published, this notebook will appear in the Azure Cosmos DB notebooks public gallery. Make sure you have removed any sensitive data or output before publishing.";

  const descriptionPara2 = `Would you like to publish and share "${FileSystemUtil.stripExtension(
    notebookName,
    "ipynb"
  )}" to the gallery?`;

  const options: ImageTypes[] = [ImageTypes.CustomImage, ImageTypes.Url];
  if (onTakeSnapshot) {
    options.push(ImageTypes.TakeScreenshot);
    if (notebookObject) {
      options.push(ImageTypes.UseFirstDisplayOutput);
    }
  }

  const thumbnailSelectorProps: IDropdownProps = {
    label: "Cover image",
    selectedKey: type,
    ariaLabel: "Cover image",
    options: options.map((value: string) => ({ text: value, key: value })),
    onChange: async (event, options) => {
      setImageSrc("");
      clearFormError();
      if (options.text === ImageTypes.TakeScreenshot) {
        onTakeSnapshot({
          aspectRatio: cardHeightToWidthRatio,
          requestId: new Date().getTime().toString(),
          type: "notebook",
          notebookContentRef,
        });
      } else if (options.text === ImageTypes.UseFirstDisplayOutput) {
        const cellIds = NotebookUtil.findCodeCellWithDisplay(notebookObject);
        if (cellIds.length > 0) {
          onTakeSnapshot({
            aspectRatio: cardHeightToWidthRatio,
            requestId: new Date().getTime().toString(),
            type: "celloutput",
            cellId: cellIds[0],
            notebookContentRef,
          });
        } else {
          firstOutputErrorHandler(new Error("Output does not exist for any of the cells."));
        }
      }
      setType(options.text);
    },
  };

  const thumbnailUrlProps: ITextFieldProps = {
    label: "Cover image url",
    ariaLabel: "Cover image url",
    required: true,
    onChange: (event, newValue) => {
      setImageSrc(newValue);
    },
  };

  const firstOutputErrorHandler = (error: Error) => {
    const formError = "Failed to capture first output";
    const formErrorDetail = `${error}`;
    const area = "PublishNotebookPaneComponent/UseFirstOutput";
    onError(formError, formErrorDetail, area);
  };

  const imageToBase64 = (file: File, updateImageSrc: (result: string) => void) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      updateImageSrc(reader.result.toString());
    };

    reader.onerror = (error) => {
      const formError = `Failed to convert ${file.name} to base64 format`;
      const formErrorDetail = `${error}`;
      const area = "PublishNotebookPaneComponent/selectImageFile";
      onError(formError, formErrorDetail, area);
    };
  };

  const renderThumbnailSelectors = (type: string) => {
    switch (type) {
      case ImageTypes.Url:
        return <TextField {...thumbnailUrlProps} />;
      case ImageTypes.CustomImage:
        return (
          <input
            id="selectImageFile"
            type="file"
            accept="image/*"
            onChange={(event) => {
              const file = event.target.files[0];
              if (file.size / 1024 ** 2 > maxImageSizeInMib) {
                event.target.value = "";
                const formError = `Failed to upload ${file.name}`;
                const formErrorDetail = `Image is larger than ${maxImageSizeInMib} MiB. Please Choose a different image.`;
                const area = "PublishNotebookPaneComponent/selectImageFile";

                onError(formError, formErrorDetail, area);
                setImageSrc("");
                return;
              } else {
                clearFormError();
              }
              imageToBase64(file, (result: string) => {
                setImageSrc(result);
              });
            }}
          />
        );
      default:
        return <></>;
    }
  };

  return (
    <div className="publishNotebookPanelContent">
      <Stack className="panelMainContent" tokens={{ childrenGap: 20 }}>
        <Stack.Item>
          <Text>{descriptionPara1}</Text>
        </Stack.Item>

        <Stack.Item>
          <Text>{descriptionPara2}</Text>
        </Stack.Item>

        <Stack.Item>
          <TextField
            label="Name"
            ariaLabel="Name"
            defaultValue={FileSystemUtil.stripExtension(notebookName, "ipynb")}
            required
            onChange={(event, newValue) => {
              const notebookName = newValue + ".ipynb";
              setNotebookName(notebookName);
            }}
          />
        </Stack.Item>

        <Stack.Item>
          <TextField
            label="Description"
            ariaLabel="Description"
            multiline
            rows={3}
            required
            onChange={(event, newValue) => {
              setNotebookDescription(newValue);
            }}
          />
        </Stack.Item>

        <Stack.Item>
          <TextField
            label="Tags"
            ariaLabel="Tags"
            placeholder="Optional tag 1, Optional tag 2"
            onChange={(event, newValue) => {
              setNotebookTags(newValue);
            }}
          />
        </Stack.Item>

        <Stack.Item>
          <Dropdown {...thumbnailSelectorProps} />
        </Stack.Item>

        <Stack.Item>{renderThumbnailSelectors(type)}</Stack.Item>

        <Stack.Item>
          <Text>Preview</Text>
        </Stack.Item>
        <Stack.Item>
          <GalleryCardComponent
            data={{
              id: undefined,
              name: notebookName,
              description: notebookDescription,
              gitSha: undefined,
              tags: notebookTags.split(","),
              author: notebookAuthor,
              thumbnailUrl: imageSrc,
              created: notebookCreatedDate,
              isSample: false,
              downloads: undefined,
              favorites: undefined,
              views: undefined,
              newCellId: undefined,
              policyViolations: undefined,
              pendingScanJobIds: undefined,
            }}
            isFavorite={undefined}
            showDownload={false}
            showDelete={false}
            onClick={() => undefined}
            onTagClick={undefined}
            onFavoriteClick={undefined}
            onUnfavoriteClick={undefined}
            onDownloadClick={undefined}
            onDeleteClick={undefined}
          />
        </Stack.Item>
      </Stack>
    </div>
  );
};
