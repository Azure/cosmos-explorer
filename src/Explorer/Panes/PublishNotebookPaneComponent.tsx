import { ITextFieldProps, Stack, Text, TextField, Dropdown, IDropdownProps } from "office-ui-fabric-react";
import * as React from "react";
import { GalleryCardComponent } from "../Controls/NotebookGallery/Cards/GalleryCardComponent";
import { FileSystemUtil } from "../Notebook/FileSystemUtil";
import "./PublishNotebookPaneComponent.less";
import Html2Canvas from "html2canvas";
import { ImmutableNotebook } from "@nteract/commutable/src";
import { NotebookUtil } from "../Notebook/NotebookUtil";

export interface PublishNotebookPaneProps {
  notebookName: string;
  notebookDescription: string;
  notebookTags: string;
  notebookAuthor: string;
  notebookCreatedDate: string;
  notebookObject: ImmutableNotebook;
  notebookParentDomElement: HTMLElement;
  onChangeName: (newValue: string) => void;
  onChangeDescription: (newValue: string) => void;
  onChangeTags: (newValue: string) => void;
  onChangeImageSrc: (newValue: string) => void;
  onError: (formError: string, formErrorDetail: string, area: string) => void;
  clearFormError: () => void;
}

interface PublishNotebookPaneState {
  type: string;
  notebookName: string;
  notebookDescription: string;
  notebookTags: string;
  imageSrc: string;
}

enum ImageTypes {
  Url = "URL",
  CustomImage = "Custom Image",
  TakeScreenshot = "Take Screenshot",
  UseFirstDisplayOutput = "Use First Display Output"
}

export class PublishNotebookPaneComponent extends React.Component<PublishNotebookPaneProps, PublishNotebookPaneState> {
  private static readonly maxImageSizeInMib = 1.5;
  private descriptionPara1: string;
  private descriptionPara2: string;
  private nameProps: ITextFieldProps;
  private descriptionProps: ITextFieldProps;
  private tagsProps: ITextFieldProps;
  private thumbnailUrlProps: ITextFieldProps;
  private thumbnailSelectorProps: IDropdownProps;
  private imageToBase64: (file: File, updateImageSrc: (result: string) => void) => void;
  private takeScreenshot: (target: HTMLElement, onError: (error: Error) => void) => void;

  constructor(props: PublishNotebookPaneProps) {
    super(props);

    this.state = {
      type: ImageTypes.Url,
      notebookName: props.notebookName,
      notebookDescription: "",
      notebookTags: "",
      imageSrc: undefined
    };

    this.imageToBase64 = (file: File, updateImageSrc: (result: string) => void) => {
      const reader = new FileReader();
      reader.readAsDataURL(file);
      reader.onload = () => {
        updateImageSrc(reader.result.toString());
      };

      const onError = this.props.onError;
      reader.onerror = error => {
        const formError = `Failed to convert ${file.name} to base64 format`;
        const formErrorDetail = `${error}`;
        const area = "PublishNotebookPaneComponent/selectImageFile";
        onError(formError, formErrorDetail, area);
      };
    };

    this.takeScreenshot = (target: HTMLElement, onError: (error: Error) => void): void => {
      const updateImageSrcWithScreenshot = (canvasUrl: string): void => {
        this.props.onChangeImageSrc(canvasUrl);
        this.setState({ imageSrc: canvasUrl });
      };

      target.scrollIntoView();
      Html2Canvas(target, {
        useCORS: true,
        allowTaint: true,
        scale: 1,
        logging: true
      })
        .then(canvas => {
          //redraw canvas to fit Card Cover Image dimensions
          const originalImageData = canvas.toDataURL();
          const requiredHeight =
            parseInt(canvas.style.width.split("px")[0]) * GalleryCardComponent.cardHeightToWidthRatio;
          canvas.height = requiredHeight;
          const context = canvas.getContext("2d");
          const image = new Image();
          image.src = originalImageData;
          image.onload = () => {
            context.drawImage(image, 0, 0);
            updateImageSrcWithScreenshot(canvas.toDataURL());
          };
        })
        .catch(error => {
          onError(error);
        });
    };

    this.descriptionPara1 =
      "This notebook has your data. Please make sure you delete any sensitive data/output before publishing.";

    this.descriptionPara2 = `Would you like to publish and share "${FileSystemUtil.stripExtension(
      this.props.notebookName,
      "ipynb"
    )}" to the gallery?`;

    this.thumbnailUrlProps = {
      label: "Cover image url",
      ariaLabel: "Cover image url",
      onChange: (event, newValue) => {
        this.props.onChangeImageSrc(newValue);
        this.setState({ imageSrc: newValue });
      }
    };

    const screenshotErrorHandler = (error: Error) => {
      const formError = "Failed to take screen shot";
      const formErrorDetail = `${error}`;
      const area = "PublishNotebookPaneComponent/takeScreenshot";
      this.props.onError(formError, formErrorDetail, area);
    };

    const firstOutputErrorHandler = (error: Error) => {
      const formError = "Failed to capture first output";
      const formErrorDetail = `${error}`;
      const area = "PublishNotebookPaneComponent/UseFirstOutput";
      this.props.onError(formError, formErrorDetail, area);
    };

    this.thumbnailSelectorProps = {
      label: "Cover image",
      defaultSelectedKey: ImageTypes.Url,
      ariaLabel: "Cover image",
      options: [
        ImageTypes.Url,
        ImageTypes.CustomImage,
        ImageTypes.TakeScreenshot,
        ImageTypes.UseFirstDisplayOutput
      ].map((value: string) => ({ text: value, key: value })),
      onChange: async (event, options) => {
        this.props.clearFormError();
        if (options.text === ImageTypes.TakeScreenshot) {
          try {
            await this.takeScreenshot(this.props.notebookParentDomElement, screenshotErrorHandler);
          } catch (error) {
            screenshotErrorHandler(error);
          }
        } else if (options.text === ImageTypes.UseFirstDisplayOutput) {
          try {
            await this.takeScreenshot(this.findFirstOutput(), firstOutputErrorHandler);
          } catch (error) {
            firstOutputErrorHandler(error);
          }
        }
        this.setState({ type: options.text });
      }
    };

    this.nameProps = {
      label: "Name",
      ariaLabel: "Name",
      defaultValue: this.props.notebookName,
      required: true,
      onChange: (event, newValue) => {
        this.props.onChangeName(newValue);
        this.setState({ notebookName: newValue });
      }
    };

    this.descriptionProps = {
      label: "Description",
      ariaLabel: "Description",
      multiline: true,
      rows: 3,
      required: true,
      onChange: (event, newValue) => {
        this.props.onChangeDescription(newValue);
        this.setState({ notebookDescription: newValue });
      }
    };

    this.tagsProps = {
      label: "Tags",
      ariaLabel: "Tags",
      placeholder: "Optional tag 1, Optional tag 2",
      onChange: (event, newValue) => {
        this.props.onChangeTags(newValue);
        this.setState({ notebookTags: newValue });
      }
    };
  }

  private renderThumbnailSelectors(type: string) {
    switch (type) {
      case ImageTypes.Url:
        return <TextField {...this.thumbnailUrlProps} />;
      case ImageTypes.CustomImage:
        return (
          <input
            id="selectImageFile"
            type="file"
            accept="image/*"
            onChange={event => {
              const file = event.target.files[0];
              if (file.size / 1024 ** 2 > PublishNotebookPaneComponent.maxImageSizeInMib) {
                event.target.value = "";
                const formError = `Failed to upload ${file.name}`;
                const formErrorDetail = `Image is larger than ${PublishNotebookPaneComponent.maxImageSizeInMib} MiB. Please Choose a different image.`;
                const area = "PublishNotebookPaneComponent/selectImageFile";

                this.props.onError(formError, formErrorDetail, area);
                this.props.onChangeImageSrc(undefined);
                this.setState({ imageSrc: undefined });
                return;
              } else {
                this.props.clearFormError();
              }
              this.imageToBase64(file, (result: string) => {
                this.props.onChangeImageSrc(result);
                this.setState({ imageSrc: result });
              });
            }}
          />
        );
      default:
        return <></>;
    }
  }

  private findFirstOutput(): HTMLElement {
    const indexOfFirstCodeCellWithDisplay = NotebookUtil.findFirstCodeCellWithDisplay(this.props.notebookObject);
    const cellOutputDomElements = this.props.notebookParentDomElement.querySelectorAll<HTMLElement>(
      ".nteract-cell-outputs"
    );
    return cellOutputDomElements[indexOfFirstCodeCellWithDisplay];
  }

  public render(): JSX.Element {
    return (
      <div className="publishNotebookPanelContent">
        <Stack className="panelMainContent" tokens={{ childrenGap: 20 }}>
          <Stack.Item>
            <Text>{this.descriptionPara1}</Text>
          </Stack.Item>

          <Stack.Item>
            <Text>{this.descriptionPara2}</Text>
          </Stack.Item>

          <Stack.Item>
            <TextField {...this.nameProps} />
          </Stack.Item>

          <Stack.Item>
            <TextField {...this.descriptionProps} />
          </Stack.Item>

          <Stack.Item>
            <TextField {...this.tagsProps} />
          </Stack.Item>

          <Stack.Item>
            <Dropdown {...this.thumbnailSelectorProps} />
          </Stack.Item>

          <Stack.Item>{this.renderThumbnailSelectors(this.state.type)}</Stack.Item>

          <Stack.Item>
            <Text>Preview</Text>
          </Stack.Item>
          <Stack.Item>
            <GalleryCardComponent
              data={{
                id: undefined,
                name: this.props.notebookName,
                description: this.state.notebookDescription,
                gitSha: undefined,
                tags: this.state.notebookTags.split(","),
                author: this.props.notebookAuthor,
                thumbnailUrl: this.state.imageSrc,
                created: this.props.notebookCreatedDate,
                isSample: false,
                downloads: 0,
                favorites: 0,
                views: 0,
                newCellId: undefined
              }}
              isFavorite={false}
              showDownload={true}
              showDelete={true}
              onClick={undefined}
              onTagClick={undefined}
              onFavoriteClick={undefined}
              onUnfavoriteClick={undefined}
              onDownloadClick={undefined}
              onDeleteClick={undefined}
            />
          </Stack.Item>
        </Stack>
      </div>
    );
  }
}
