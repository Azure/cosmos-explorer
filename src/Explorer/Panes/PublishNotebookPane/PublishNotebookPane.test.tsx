import { shallow } from "enzyme";
import React from "react";
import { PublishNotebookPaneComponent, PublishNotebookPaneProps } from "./PublishNotebookPaneComponent";

describe("PublishNotebookPaneComponent", () => {
  it("renders", () => {
    const props: PublishNotebookPaneProps = {
      notebookName: "SampleNotebook.ipynb",
      notebookDescription: "sample description",
      notebookTags: "tag1, tag2",
      imageSrc: "https://i.ytimg.com/vi/E_lByLdKeKY/maxresdefault.jpg",
      notebookAuthor: "CosmosDB",
      notebookCreatedDate: "2020-07-17T00:00:00Z",
      notebookObject: undefined,
      notebookContentRef: undefined,
      setNotebookName: undefined,
      setNotebookDescription: undefined,
      setNotebookTags: undefined,
      setImageSrc: undefined,
      onError: undefined,
      clearFormError: undefined,
      onTakeSnapshot: undefined,
    };

    const wrapper = shallow(<PublishNotebookPaneComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
