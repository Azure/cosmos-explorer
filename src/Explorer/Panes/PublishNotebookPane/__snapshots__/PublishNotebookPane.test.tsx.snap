// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublishNotebookPaneComponent renders 1`] = `
<div
  className="publishNotebookPanelContent"
>
  <Stack
    className="panelMainContent"
    tokens={
      Object {
        "childrenGap": 20,
      }
    }
  >
    <StackItem>
      <Text>
        When published, this notebook will appear in the Azure Cosmos DB notebooks public gallery. Make sure you have removed any sensitive data or output before publishing.
      </Text>
    </StackItem>
    <StackItem>
      <Text>
        Would you like to publish and share "SampleNotebook" to the gallery?
      </Text>
    </StackItem>
    <StackItem>
      <StyledTextFieldBase
        ariaLabel="Name"
        defaultValue="SampleNotebook"
        label="Name"
        onChange={[Function]}
        required={true}
      />
    </StackItem>
    <StackItem>
      <StyledTextFieldBase
        ariaLabel="Description"
        label="Description"
        multiline={true}
        onChange={[Function]}
        required={true}
        rows={3}
      />
    </StackItem>
    <StackItem>
      <StyledTextFieldBase
        ariaLabel="Tags"
        label="Tags"
        onChange={[Function]}
        placeholder="Optional tag 1, Optional tag 2"
      />
    </StackItem>
    <StackItem>
      <Dropdown
        ariaLabel="Cover image"
        label="Cover image"
        onChange={[Function]}
        options={
          Array [
            Object {
              "key": "Custom Image",
              "text": "Custom Image",
            },
            Object {
              "key": "URL",
              "text": "URL",
            },
          ]
        }
        selectedKey="Custom Image"
      />
    </StackItem>
    <StackItem>
      <input
        accept="image/*"
        id="selectImageFile"
        onChange={[Function]}
        type="file"
      />
    </StackItem>
    <StackItem>
      <Text>
        Preview
      </Text>
    </StackItem>
    <StackItem>
      <GalleryCardComponent
        data={
          Object {
            "author": "CosmosDB",
            "created": "2020-07-17T00:00:00Z",
            "description": "sample description",
            "downloads": undefined,
            "favorites": undefined,
            "gitSha": undefined,
            "id": undefined,
            "isSample": false,
            "name": "SampleNotebook.ipynb",
            "newCellId": undefined,
            "pendingScanJobIds": undefined,
            "policyViolations": undefined,
            "tags": Array [
              "tag1",
              " tag2",
            ],
            "thumbnailUrl": "https://i.ytimg.com/vi/E_lByLdKeKY/maxresdefault.jpg",
            "views": undefined,
          }
        }
        onClick={[Function]}
        showDelete={false}
        showDownload={false}
      />
    </StackItem>
  </Stack>
</div>
`;
